lib/pbench/test/unit/server/database/__init__.py

```diff
@@ -0,0 +1,351 @@
+"""The SQLAlchemy mock infrastructure here is not completely generalized, but
+is currently sufficient to support the ServerConfig and Audit DB unit tests.
+"""
+
+import operator
+from typing import Optional
+
+from sqlalchemy import Column
+from sqlalchemy.util import immutabledict
+
+from pbench.server.database.database import Database
+
+
+def dumpem(object: object, depth: int = 0, max_depth: int = 20):
+    """
+    Useful recursive function to help debug SQLAlchemy's deep and mostly not
+    documented class hierarchy! NOTE: this can generate an enormous amount of
+    output and should be used carefully. While simple in principle, it's worth
+    archiving for future use.
+
+    The recursion depth is limited to 20 by default. That's arbitrary, but we
+    don't want to overflow.
+
+    Args:
+        object: Any Python object type
+        depth: indentation levels (two spaces)
+        max_depth: maximum recursion depth
+    """
+    if depth > max_depth:
+        return
+    indent = " " * (depth * 2)
```

Comment:
Alternately, you could double the literal instead of doubling the multiplier:
```
    indent = "  " * depth
```

Reply:
Mostly I thought `* 2` was easier to grasp visually than `" "` vs `"  "`; in some fonts the doubled space might be hard to immediately pick out. On the other hand I don't see how this distinction matters at all.

(Interestingly, even inside the backticks, GitHub's markdown seems to have collapsed the two spaces in the second instance in this comment: perhaps another argument in favor of keeping the code as is. 😏 )

Reply:
I was thinking from the point of view of
```
    one_indent = "  "            # This is two spaces, but it could be one hyphen or anything else
    indent = one_indent * depth  # Indent once for each depth of call
```